Ensured text items are sorted when added to QComboBox.

chris2oph opened the pull request:
This change allows entries added to a `QComboBox` to be sorted alphabetically. This change ensures the test: **ChoiceTestCase - Sort** now passes, mostly...

There is one remaining issue with this test, and I believe there may be an issue with this test. There is a list of short strings sorted and compared in this test, and the final check expects the string "a" to have been sorted as the first item. This string actually ends up as the third item in the list, and I believe this is correct.

**Expected Sort Order**
a
AAA
Aaa
aaa
aaab
aab
aba

**Actual Sort Order**
AAA
Aaa
a
aaa
aaab
aab
aba

I'd like to update the test to expect the string "a" as the third item rather than the first, however I would like to know if anybody has any issues with this before I take this action.

chris2oph committed:
Ensured text items are sorted when added to QComboBox.

vadz commented:
Sorry, I don't understand how does this work: shouldn't we call `sort()` only for sorted comboboxes?

As for the sort order, it looks that Qt might be using case-sensitive sorting which really doesn't make much sense for user-oriented stuff. If it can be persuaded to sort case-insensitively, we should do this. Otherwise the test will need to be tweaked specifically for it, which is not ideal, but I don't see any other possibility -- i.e. we're not going to change the existing behaviour.

chris2oph committed:
Ensured that a QComboBox is only sorted when the box is created with the wxCB_SORT flag.

chris2oph commented:
I have addressed one of your comments @vadz; specifically I have added changes so that we only sort the items in a combo box if the combo box was created with the `wxCB_SORT` flag.

As for your other comment, I would like to verify that WX sorts the same list of items in the same manner prior to updating the tests. If the WX version of the sort is different to QT then we will be left with the option of attempting to write a specific sort so that QT matches WX or simply having to update the test.

vadz commented:
Sorry, I'm not sure what exactly would you like to verify? wxWidgets doesn't actually sort anything itself if the platform supports it directly (which is the case for MSW with its `CB_SORT` native combobox flag, for example). So it may well not sort the items in the same way under different platforms -- but this is fine and actually better than sorting the items in a way which is unusual for the current platform users.

BTW, #780 is somewhat related and might be useful.

chris2oph commented:
There may have been a misunderstanding here @vadz . My original aim with this change was to tackle the specific test case: **ChoiceTestCase - Sort**. To ensure this test case passes it was necessary to apply a sort to `QComboBox` (see my first commit for this change). After I had done a pull request for this change, you commented: "_shouldn't we call sort() only for sorted comboboxes?_". To which I took that to mean my initial change was applying the sort for _every_ `QComboBox`. So, I applied a further change (see my second commit for this change) that would store the style of a combo box upon creation; this would allow us to check the style when adding items to the combo box and only sort when appropriate.

The thing I needed to verify (I intend to do this myself) is whether or not WX and QT sort the list of strings in a different fashion. If they do, then the specified test will not succeed and this would entail either a change of the test, or a specific sort function written for QT (I'd like to avoid this!). If WX sorts these strings the same way as QT, then the test is simply wrong and needs to be updated.

Apologies for any confusion, if you disagree with anything I have said or have any other comments I would be happy to hear it. Once I have figured out how WX sorts these strings I will add a further comment.

vadz commented on include/wx/qt/choice.h:
```diff
@@ -72,6 +72,7 @@ class WXDLLIMPEXP_CORE wxChoice : public wxChoiceBase
     virtual void DoDeleteOneItem(unsigned int pos);
 
     QComboBox *m_qtComboBox;
+    long m_style;
```
This is unnecessary, the style is already stored in `wxWindowBase::m_windowStyle`, so it would just duplicate it.

vadz commented on src/qt/choice.cpp:
```diff
@@ -148,6 +149,10 @@ int wxChoice::DoInsertItems(const wxArrayStringsAdapter & items,
 {
     m_qtComboBox->insertItem(pos, wxQtConvertString(item));
+
+    if (m_style == wxCB_SORT)
```
```suggestion
    if ( IsSorted() )
```
(look at its implementation in `wx/choice.h` to see why/how does it work)

vadz commented:
I'm fine with applying this PR after changing it as indicated, thanks.

As for the other platforms, yes, they can sort the items differently, although apparently GTK and MSW versions agree on the sort order in this particular case (at least the test passes for them). And, unless we have something like #780, we will indeed need to change the test to accommodate the platforms using a different sort order -- I don't see what else could we do.

chris2oph committed:
Removed unnecessary member variable 'style' and utilised existing functionality to check for a sorted combo box.
